File: rust/src/domains/bulk.rs
// Copyright since 2025 Mifos Initiative
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

use crate::adapter::FineractAdapter;
use rmcp::schemars;
use rmcp::{ErrorData as McpError, model::{CallToolResult, Content}};
use serde_json::json;
use futures::future::join_all;
use chrono::Local;

pub fn to_result(val: serde_json::Value) -> Result<CallToolResult, McpError> { Ok(CallToolResult::success(vec![Content::text(serde_json::to_string_pretty(&val).unwrap_or_default())])) }
pub fn to_err(err: anyhow::Error) -> McpError { McpError::internal_error(err.to_string(), None) }
pub fn today() -> String { Local::now().format("%d %B %Y").to_string() }


#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkSearchClientsReq { pub name_queries: Vec<String> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkLoanStatusReq { pub loan_ids: Vec<i64> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct DisburseItem { pub loan_id: i64, pub amount_exactly_from_user_prompt: f64, pub transaction_date: String }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkDisburseReq { pub disbursements: Vec<DisburseItem> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct RepaymentItem { pub loan_id: i64, pub amount_exactly_from_user_prompt: f64, pub transaction_date: String, pub receipt_number: Option<String> }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkRepaymentsReq { pub repayments: Vec<RepaymentItem> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkActivateReq { pub client_ids: Vec<i64> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkSavingsBalancesReq { pub savings_ids: Vec<i64> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct ApplyFeeItem { pub client_id: i64, pub charge_id: i64, pub amount_exactly_from_user_prompt: f64 }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkApplyFeesReq { pub fees: Vec<ApplyFeeItem> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct CloseAccountItem { pub savings_id: i64, pub closure_reason_id: i64 }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkCloseAccountsReq { pub closures: Vec<CloseAccountItem> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct CreateSavingsItem { pub client_id: i64, pub product_id: Option<i64> }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkCreateSavingsReq { pub accounts: Vec<CreateSavingsItem> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkApproveActivateSavingsReq { pub savings_ids: Vec<i64> }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct DepositItem { pub savings_id: i64, pub amount_exactly_from_user_prompt: f64 }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct BulkDepositSavingsReq { pub deposits: Vec<DepositItem> }



pub async fn bulk_search_clients(adapter: &FineractAdapter, req: BulkSearchClientsReq) -> Result<CallToolResult, McpError> {
    let futures = req.name_queries.into_iter().map(|q| {
        let adapter_ref = adapter.clone();
        async move {
            let qs: [(&str, &str); 3] = [("query", &q), ("resource", "clients"), ("exactMatch", "false")];
            let res = adapter_ref.execute_get("search", Some(&qs)).await;
            json!({"query": q, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    let results = join_all(futures).await;
    to_result(json!(results))
}


pub async fn bulk_get_loan_status(adapter: &FineractAdapter, req: BulkLoanStatusReq) -> Result<CallToolResult, McpError> {
    let futures = req.loan_ids.into_iter().map(|id| {
        let adapter_ref = adapter.clone();
        async move {
            let res = adapter_ref.execute_get(&format!("loans/{}", id), None).await;
            json!({"loanId": id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_disburse_loans(adapter: &FineractAdapter, req: BulkDisburseReq) -> Result<CallToolResult, McpError> {
    let futures = req.disbursements.into_iter().map(|item| {
        let adapter_ref = adapter.clone();
        async move {
            let payload = json!({ "actualDisbursementDate": item.transaction_date, "transactionAmount": item.amount_exactly_from_user_prompt, "dateFormat": "dd MMMM yyyy", "locale": "en" });
            let res = adapter_ref.execute_post(&format!("loans/{}?command=disburse", item.loan_id), &payload).await;
            json!({"loanId": item.loan_id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_make_repayments(adapter: &FineractAdapter, req: BulkRepaymentsReq) -> Result<CallToolResult, McpError> {
    let futures = req.repayments.into_iter().map(|item| {
        let adapter_ref = adapter.clone();
        async move {
            let mut payload = json!({ "transactionDate": item.transaction_date, "transactionAmount": item.amount_exactly_from_user_prompt, "dateFormat": "dd MMMM yyyy", "locale": "en" });
            if let Some(r) = item.receipt_number { payload["receiptNumber"] = json!(r); }
            let res = adapter_ref.execute_post(&format!("loans/{}/transactions?command=repayment", item.loan_id), &payload).await;
            json!({"loanId": item.loan_id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_activate_clients(adapter: &FineractAdapter, req: BulkActivateReq) -> Result<CallToolResult, McpError> {
    let date = today();
    let futures = req.client_ids.into_iter().map(|id| {
        let adapter_ref = adapter.clone();
        let date_ref = date.clone();
        async move {
            let payload = json!({ "activationDate": date_ref, "dateFormat": "dd MMMM yyyy", "locale": "en" });
            let res = adapter_ref.execute_post(&format!("clients/{}?command=activate", id), &payload).await;
            json!({"clientId": id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_get_savings_balances(adapter: &FineractAdapter, req: BulkSavingsBalancesReq) -> Result<CallToolResult, McpError> {
    let futures = req.savings_ids.into_iter().map(|id| {
        let adapter_ref = adapter.clone();
        async move {
            let res = adapter_ref.execute_get(&format!("savingsaccounts/{}", id), None).await;
            json!({"savingsId": id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_apply_fees(adapter: &FineractAdapter, req: BulkApplyFeesReq) -> Result<CallToolResult, McpError> {
    let date = today();
    let futures = req.fees.into_iter().map(|item| {
        let adapter_ref = adapter.clone();
        let date_ref = date.clone();
        async move {
            let payload = json!({ "chargeId": item.charge_id, "amount": item.amount_exactly_from_user_prompt, "dueDate": date_ref, "dateFormat": "dd MMMM yyyy", "locale": "en" });
            let res = adapter_ref.execute_post(&format!("clients/{}/charges", item.client_id), &payload).await;
            json!({"clientId": item.client_id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_close_accounts(adapter: &FineractAdapter, req: BulkCloseAccountsReq) -> Result<CallToolResult, McpError> {
    let date = today();
    let futures = req.closures.into_iter().map(|item| {
        let adapter_ref = adapter.clone();
        let date_ref = date.clone();
        async move {
            let payload = json!({ "closedOnDate": date_ref, "closureReasonId": item.closure_reason_id, "dateFormat": "dd MMMM yyyy", "locale": "en" });
            let res = adapter_ref.execute_post(&format!("savingsaccounts/{}?command=close", item.savings_id), &payload).await;
            json!({"savingsId": item.savings_id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}

pub async fn bulk_create_savings_accounts(adapter: &FineractAdapter, req: BulkCreateSavingsReq) -> Result<CallToolResult, McpError> {
    let date = today();
    let futures = req.accounts.into_iter().map(|item| {
        let adapter_ref = adapter.clone();
        let date_ref = date.clone();
        async move {
            let payload = json!({"clientId": item.client_id, "productId": item.product_id.unwrap_or(1), "submittedOnDate": date_ref, "dateFormat": "dd MMMM yyyy", "locale": "en"});
            let res = adapter_ref.execute_post("savingsaccounts", &payload).await;
            json!({"clientId": item.client_id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_approve_and_activate_savings(adapter: &FineractAdapter, req: BulkApproveActivateSavingsReq) -> Result<CallToolResult, McpError> {
    let date = today();
    let futures = req.savings_ids.into_iter().map(|id| {
        let adapter_ref = adapter.clone();
        let date_ref = date.clone();
        async move {
            let approve_payload = json!({"dateFormat": "dd MMMM yyyy", "locale": "en", "approvedOnDate": date_ref});
            let _ = adapter_ref.execute_post(&format!("savingsaccounts/{}?command=approve", id), &approve_payload).await;
            let activate_payload = json!({"dateFormat": "dd MMMM yyyy", "locale": "en", "activatedOnDate": date_ref});
            let res = adapter_ref.execute_post(&format!("savingsaccounts/{}?command=activate", id), &activate_payload).await;
            json!({"savingsId": id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}


pub async fn bulk_deposit_savings(adapter: &FineractAdapter, req: BulkDepositSavingsReq) -> Result<CallToolResult, McpError> {
    let date = today();
    let futures = req.deposits.into_iter().map(|item| {
        let adapter_ref = adapter.clone();
        let date_ref = date.clone();
        async move {
            let payload = json!({"transactionDate": date_ref, "transactionAmount": item.amount_exactly_from_user_prompt, "dateFormat": "dd MMMM yyyy", "locale": "en", "paymentTypeId": 1});
            let res = adapter_ref.execute_post(&format!("savingsaccounts/{}/transactions?command=deposit", item.savings_id), &payload).await;
            json!({"savingsId": item.savings_id, "result": res.unwrap_or_else(|e| json!({"error": e.to_string()}))})
        }
    });
    to_result(json!(join_all(futures).await))
}

File: go/tools/bulk.go
// Copyright since 2025 Mifos Initiative
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
)

type BulkDef struct {
	Name        string
	Description string
	Method      string
	EndpointURL string
}

func (r *Registry) RegisterBulkTools() {
	bulkDefs := []BulkDef{
		{"bulk_search_clients", "Search multiple client names concurrently", "GET", "search"},
		{"bulk_get_loan_status", "Get loan details for multiple loans concurrently", "GET", "loans/%v"},
		{"bulk_disburse_loans", "Disburse multiple loans concurrently", "POST", "loans/%v?command=disburse"},
		{"bulk_make_repayments", "Make loop repayments concurrently", "POST", "loans/%v/transactions?command=repayment"},
		{"bulk_activate_clients", "Activate multiple clients concurrently", "POST", "clients/%v?command=activate"},
		{"bulk_get_savings_balances", "Get savings details concurrently", "GET", "savingsaccounts/%v"},
		{"bulk_apply_fees", "Apply fees to multiple loans simultaneously", "POST", "loans/%v/charges"},
		{"bulk_close_accounts", "Close multiple savings accounts concurrently", "POST", "savingsaccounts/%v?command=close"},
		{"bulk_create_savings_accounts", "Create multiple savings accounts concurrently", "POST", "savingsaccounts"},
		{"bulk_approve_and_activate_savings", "Approve and activate multiple savings concurrently", "POST", "savingsaccounts/%v?command=approve"},
		{"bulk_deposit_savings", "Deposit into multiple savings concurrently", "POST", "savingsaccounts/%v/transactions?command=deposit"},
	}

	for _, d := range bulkDefs {
		tool := mcp.NewTool(d.Name,
			mcp.WithDescription(d.Description),
			mcp.WithString("requests", mcp.Required(), mcp.Description("A JSON array of parameter block objects corresponding to the single operation parameters.")),
		)

		def := d
		r.Server.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {

			args, ok := req.Params.Arguments.(map[string]interface{})
			if !ok {
				return mcp.NewToolResultError("Arguments must be a valid JSON object"), nil
			}

			requestsRaw, ok := args["requests"].(string)
			if !ok {
				return mcp.NewToolResultError("Missing 'requests' parameter as stringified JSON array"), nil
			}

			var payloadArray []map[string]interface{}
			if err := json.Unmarshal([]byte(requestsRaw), &payloadArray); err != nil {
				return mcp.NewToolResultError("Failed to parse 'requests' array JSON: " + err.Error()), nil
			}

			var wg sync.WaitGroup
			results := make([]string, len(payloadArray))
			errorsFound := false

			for i, payload := range payloadArray {
				wg.Add(1)
				go func(idx int, p map[string]interface{}) {
					defer wg.Done()

					endpoint := def.EndpointURL
					queryParams := make(map[string]string)
					bodyParams := make(map[string]interface{})

					if strings.Contains(endpoint, "%v") {
						idVal := p["clientId"]
						if idVal == nil {
							idVal = p["loanId"]
						}
						if idVal == nil {
							idVal = p["savingsId"]
						}
						if idVal == nil {
							idVal = p["groupId"]
						}

						if idVal != nil {
							endpoint = strings.Replace(endpoint, "%v", fmt.Sprintf("%v", idVal), 1)
						}
					}

					if def.Method == "GET" {
						for k, v := range p {
							queryParams[k] = fmt.Sprintf("%v", v)
						}
					} else {
						for k, v := range p {
							if k == "clientId" || k == "loanId" || k == "savingsId" || k == "groupId" {
								continue
							}
							bodyParams[k] = v
						}
					}

					respData, err := r.Fineract.DoRequest(def.Method, endpoint, bodyParams, queryParams)
					if err != nil {
						results[idx] = fmt.Sprintf("Error on index %d: %v", idx, err)
						errorsFound = true
					} else {
						results[idx] = string(respData)
					}
				}(i, payload)
			}

			wg.Wait()

			finalOutput := "[" + strings.Join(results, ",") + "]"
			if errorsFound {
				return mcp.NewToolResultError("SUCCESS: Executed " + def.Name + ". Bulk execution completed with some errors: " + finalOutput), nil
			}
			return mcp.NewToolResultText("SUCCESS: Executed " + def.Name + ". Result: " + finalOutput), nil
		})
	}
}

func (r *Registry) RegisterNativeBulkTools() {
	defs := r.GetNativeBulkToolDefs()
	r.registerTools(defs)
}

func (r *Registry) GetNativeBulkToolDefs() []BaseToolDef {
	return []BaseToolDef{
		{
			Name:        "bulk_get_client_profiles",
			Description: "Fetch full details for multiple Client IDs in parallel (Go-Native)",
			Handler: func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				args := req.GetArguments()
				clientIdsRaw, ok := args["clientIds"].([]any)
				if !ok {
					return mcp.NewToolResultError("clientIds must be an array of numbers"), nil
				}

				var wg sync.WaitGroup
				results := make([]string, len(clientIdsRaw))
				var mu sync.Mutex

				for i, idRaw := range clientIdsRaw {
					wg.Add(1)
					go func(idx int, id interface{}) {
						defer wg.Done()
						endpoint := fmt.Sprintf("clients/%v", id)
						respData, err := r.Fineract.DoRequest("GET", endpoint, nil, nil)
						
						mu.Lock()
						if err != nil {
							results[idx] = fmt.Sprintf(`{"id": %v, "error": "%v"}`, id, err)
						} else {
							results[idx] = string(respData)
						}
						mu.Unlock()
					}(i, idRaw)
				}
				wg.Wait()
				return mcp.NewToolResultText("[" + strings.Join(results, ",") + "]"), nil
			},
			Params: []ToolParam{
				{Name: "clientIds", Description: "Array of Client IDs to fetch", Type: "array", Required: true},
			},
		},
		{
			Name:        "bulk_get_loan_summaries",
			Description: "Fetch status and details for multiple Loan IDs in parallel (Go-Native)",
			Handler: func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				args := req.GetArguments()
				loanIdsRaw, ok := args["loanIds"].([]any)
				if !ok {
					return mcp.NewToolResultError("loanIds must be an array of numbers"), nil
				}

				var wg sync.WaitGroup
				results := make([]string, len(loanIdsRaw))
				var mu sync.Mutex

				for i, idRaw := range loanIdsRaw {
					wg.Add(1)
					go func(idx int, id interface{}) {
						defer wg.Done()
						endpoint := fmt.Sprintf("loans/%v", id)
						respData, err := r.Fineract.DoRequest("GET", endpoint, nil, nil)
						
						mu.Lock()
						if err != nil {
							results[idx] = fmt.Sprintf(`{"id": %v, "error": "%v"}`, id, err)
						} else {
							results[idx] = string(respData)
						}
						mu.Unlock()
					}(i, idRaw)
				}
				wg.Wait()
				return mcp.NewToolResultText("[" + strings.Join(results, ",") + "]"), nil
			},
			Params: []ToolParam{
				{Name: "loanIds", Description: "Array of Loan IDs to fetch", Type: "array", Required: true},
			},
		},
		{
			Name:        "bulk_search_clients_parallel",
			Description: "Search for multiple name fragments concurrently (Go-Native)",
			Handler: func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				args := req.GetArguments()
				queriesRaw, ok := args["queries"].([]any)
				if !ok {
					return mcp.NewToolResultError("queries must be an array of strings"), nil
				}

				var wg sync.WaitGroup
				results := make([]string, len(queriesRaw))
				var mu sync.Mutex

				for i, queryRaw := range queriesRaw {
					wg.Add(1)
					go func(idx int, q interface{}) {
						defer wg.Done()
						queryParams := map[string]string{"displayName": fmt.Sprintf("%v", q)}
						respData, err := r.Fineract.DoRequest("GET", "clients", nil, queryParams)
						
						mu.Lock()
						if err != nil {
							results[idx] = fmt.Sprintf(`{"query": "%v", "error": "%v"}`, q, err)
						} else {
							results[idx] = string(respData)
						}
						mu.Unlock()
					}(i, queryRaw)
				}
				wg.Wait()
				return mcp.NewToolResultText("[" + strings.Join(results, ",") + "]"), nil
			},
			Params: []ToolParam{
				{Name: "queries", Description: "Array of search strings", Type: "array", Required: true},
			},
		},
		{
			Name:        "bulk_apply_charges_to_batch",
			Description: "Apply a single charge to multiple clients concurrently (Go-Native)",
			Handler: func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				args := req.GetArguments()
				clientIdsRaw, ok := args["clientIds"].([]any)
				chargeId, ok2 := args["chargeId"].(float64)
				if !ok || !ok2 {
					return mcp.NewToolResultError("Missing clientIds (array) or chargeId (number)"), nil
				}

				var wg sync.WaitGroup
				results := make([]string, len(clientIdsRaw))
				var mu sync.Mutex

				body := map[string]interface{}{
					"chargeId": chargeId,
					"amount":   args["amount"],
					"dueDate":  args["dueDate"],
				}

				for i, idRaw := range clientIdsRaw {
					wg.Add(1)
					go func(idx int, id interface{}) {
						defer wg.Done()
						endpoint := fmt.Sprintf("clients/%v/charges", id)
						respData, err := r.Fineract.DoRequest("POST", endpoint, body, nil)
						
						mu.Lock()
						if err != nil {
							results[idx] = fmt.Sprintf(`{"clientId": %v, "error": "%v"}`, id, err)
						} else {
							results[idx] = string(respData)
						}
						mu.Unlock()
					}(i, idRaw)
				}
				wg.Wait()
				return mcp.NewToolResultText("[" + strings.Join(results, ",") + "]"), nil
			},
			Params: []ToolParam{
				{Name: "clientIds", Description: "Array of Client IDs", Type: "array", Required: true},
				{Name: "chargeId", Description: "ID of the charge to apply", Type: "number", Required: true},
				{Name: "amount", Description: "Optional amount override", Type: "number", Required: false},
				{Name: "dueDate", Description: "Optional due date", Type: "string", Required: false},
			},
		},
		{
			Name:        "bulk_make_repayments_native",
			Description: "Submit multiple parallel loan repayments. MISSION CRITICAL: Use the exact amount from the human prompt. Do NOT adjust based on outstanding balances.",
			Handler: func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				args := req.GetArguments()
				repaymentsRaw, ok := args["repayments"].([]any)
				if !ok {
					return mcp.NewToolResultError("repayments must be an array of objects"), nil
				}

				var wg sync.WaitGroup
				results := make([]string, len(repaymentsRaw))
				var mu sync.Mutex

				sem := make(chan struct{}, 10)
				today := time.Now().Format("2006-01-02")

				for i, repRaw := range repaymentsRaw {
					wg.Add(1)
					rep, ok := repRaw.(map[string]interface{})
					if !ok {
						results[i] = `{"error": "Invalid repayment object"}`
						wg.Done()
						continue
					}

					go func(idx int, p map[string]interface{}) {
						defer wg.Done()
						sem <- struct{}{}
						defer func() { <-sem }()

						loanId := p["loanId"]
						amountRaw := p["amount"]
						date := p["transactionDate"]

						if date == nil || date == "" {
							date = today
						}

						amount, err := FormatAmount(amountRaw)
						if err != nil {
							mu.Lock()
							results[idx] = fmt.Sprintf(`{"loanId": %v, "error": "Invalid amount: %v"}`, loanId, err)
							mu.Unlock()
							return
						}

						body := map[string]interface{}{
							"transactionDate": date,
							"transactionAmount": amount,
							"dateFormat":      "yyyy-MM-dd",
							"locale":          "en",
						}
						
						endpoint := fmt.Sprintf("loans/%v/transactions?command=repayment", loanId)
						respData, err := r.Fineract.DoRequest("POST", endpoint, body, nil)
						
						mu.Lock()
						if err != nil {
							results[idx] = fmt.Sprintf(`{"loanId": %v, "error": "%v"}`, loanId, err)
						} else {
							results[idx] = string(respData)
						}
						mu.Unlock()
					}(i, rep)
				}
				wg.Wait()
				return mcp.NewToolResultText("SUCCESS: Executed bulk_make_repayments_native. Results: [" + strings.Join(results, ",") + "]"), nil
			},
			Params: []ToolParam{
				{Name: "repayments", Description: "Array of objects. MISSION CRITICAL: Each 'amount' field MUST be the exact numeric value from the human prompt. IGNORE backend due amounts.", Type: "array", Required: true},
			},
		},
	}
}
